controllers/metal3.io/bmceventsubscription_controller.go

```diff
@@ -0,0 +1,280 @@
+
+			return ctrl.Result{RequeueAfter: subscriptionRetryDelay}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrap(err, "could not load host data")
+	}
+
+	err = r.addFinalizer(ctx, subscription)
+
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed add finalizer")
+	}
+
+	prov, ready, err := r.getProvisioner(request, host)
+
+	if err != nil {
+		return ctrl.Result{}, errors.Wrap(err, "failed to create provisioner")
+	}
+
+	if !ready {
+		reqLogger.Info("provisioner is not ready", "RequeueAfter:", provisionerNotReadyRetryDelay)
+		return ctrl.Result{RequeueAfter: provisionerNotReadyRetryDelay}, nil
+	}
+
+	if subscription.DeletionTimestamp.IsZero() {
+		// Not being deleted
+		if err := r.createSubscription(ctx, prov, subscription); err != nil {
+			subscription.Status.Error = fmt.Sprintf("failed to create a subscription")
+			err := r.Status().Update(ctx, subscription)
+			if err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to update subscription status")
+			}
+			return ctrl.Result{}, errors.Wrap(err, "failed to create subscription")
+		}
+	} else {
+		// Being deleted
+		if err := r.deleteSubscription(ctx, prov, subscription); err != nil {
+			subscription.Status.Error = fmt.Sprintf("failed to delete subscription")
+			err := r.Status().Update(ctx, subscription)
+			if err != nil {
+				return ctrl.Result{}, errors.Wrap(err, "failed to update subscription status")
+			}
+			return ctrl.Result{}, errors.Wrap(err, "failed to delete subscription")
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	return
+
+}
+
+func (r *BMCEventSubscriptionReconciler) addFinalizer(ctx context.Context, subscription *metal3v1alpha1.BMCEventSubscription) error {
+	reqLogger := r.Log.WithName("bmceventsubscription")
+
+	// Add a finalizer to newly created objects.
+	if subscription.DeletionTimestamp.IsZero() && !subscriptionHasFinalizer(subscription) {
+		reqLogger.Info(
+			"adding finalizer",
+			"existingFinalizers", subscription.Finalizers,
+			"newValue", metal3v1alpha1.BMCEventSubscriptionFinalizer,
+		)
+		subscription.Finalizers = append(subscription.Finalizers,
+			metal3v1alpha1.BMCEventSubscriptionFinalizer)
+		err := r.Update(ctx, subscription)
+		if err != nil {
+			return errors.Wrap(err, "failed to add finalizer")
+		}
+		return nil
+	}
+
+	return nil
+}
+
+func (r *BMCEventSubscriptionReconciler) createSubscription(ctx context.Context, prov provisioner.Provisioner, subscription *metal3v1alpha1.BMCEventSubscription) error {
+	reqLogger := r.Log.WithName("bmceventsubscription")
+
+	if subscription.Status.SubscriptionID != "" {
+		reqLogger.Info("subscription already exists")
+		return nil
+	}
+
+	headers, err := r.getHTTPHeaders(*subscription)
+
+	if err != nil {
+		reqLogger.Error(err, "failed to get http headers")
+		subscription.Status.Error = fmt.Sprintf("failed to retrieve HTTP headers secret")
+		updateErr := r.Status().Update(ctx, subscription)
+		if updateErr != nil {
```

Comment:
This will get shadowed in the main loop [call](https://github.com/metal3-io/baremetal-operator/pull/1014/files#diff-2e58d31f13465b9d9eae85cfdf77bfa0579b6d3cd553b1fc7c37162c4e1f6e22R121)

Reply:
Note: updating the status in case of a permanent error (ie wrong http header) will cause an hot loop. To avoid it, please add an update predicate function like the [one](https://github.com/metal3-io/baremetal-operator/blob/3f19cf4344fed69def3a167a86a1375405840cac/controllers/metal3.io/baremetalhost_controller.go#L1557-L1579) used in the BMH controller